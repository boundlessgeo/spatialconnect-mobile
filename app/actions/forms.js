'use strict';
import * as sc from 'spatialconnect/native';

export const loadForms = () => {
  return dispatch => {
    sc.forms().subscribe(data => {
      dispatch({
        type: 'ADD_FORM_LIST',
        forms: data.forms
      });
    });
  };
};