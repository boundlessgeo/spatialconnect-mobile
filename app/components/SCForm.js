'use strict';
import React, { Component, PropTypes } from 'react';
import {
  ScrollView,
  StyleSheet,
  Text,
  View
} from 'react-native';
import { Actions } from 'react-native-router-flux';
import Button from 'react-native-button';
import transform from 'tcomb-json-schema';
import tcomb from 'tcomb-form-native';
import palette from '../style/palette';
import { buttonStyles } from '../style/style';
import scformschema from 'spatialconnect-form-schema/native';
import * as sc from 'spatialconnect/native';

tcomb.form.Form.stylesheet.textbox.normal.backgroundColor = '#ffffff';
transform.registerType('date', tcomb.Date);
transform.registerType('time', tcomb.Date);

let Form = tcomb.form.Form;

class SCForm extends Component {

  constructor(props) {
    super(props);
    this.state = {
      location: null,
      value: {}
    };
  }

  saveForm(formData) {
    navigator.geolocation.getCurrentPosition(position => {
      let gj = {
        geometry: {
          type: 'Point',
          coordinates: [
            position.coords.longitude,
            position.coords.latitude
          ]
        },
        properties: formData
      };
      let f = sc.geometry('DEFAULT_STORE', this.props.formInfo.layer_name, gj);
      sc.createFeature$(f.serialize()).first().subscribe((data) => {
        Actions.formSubmitted({ feature: data.payload });
      });
    }, (error) => {
      let f = sc.spatialFeature('DEFAULT_STORE', this.props.formInfo.layer_name, formData);
      sc.createFeature$(f.serialize()).first().subscribe((data) => {
        Actions.formSubmitted({ feature: data.payload });
      });
    },
      {enableHighAccuracy: true, timeout: 20000, maximumAge: 1000}
    );
  }

  onPress() {
    var value = this.refs.form.getValue();
    if (value) {
      this.saveForm(value);
    }
  }

  componentWillMount() {
    let { schema, options } = scformschema.translate(this.props.formInfo);
    let initialValues = {};
    for (let prop in schema.properties) {
      if (schema.properties[prop].hasOwnProperty('initialValue')) {
        initialValues[prop] = schema.properties[prop].initialValue;
      }
    }
    this.setState({value: initialValues});
    this.TcombType = transform(schema);
    this.options = options;
  }

  onChange(value) {
    this.setState({value});
  }

  render() {
    return (
      <View style={styles.container}>
        <ScrollView style={styles.scrollView}>
          <View style={styles.formName}>
            <Text style={styles.formNameText}>{this.props.formInfo.display_name}</Text>
          </View>
          <View style={styles.form}>
            <Form
              ref="form"
              value={this.state.value}
              type={this.TcombType}
              options={this.options}
              onChange={this.onChange.bind(this)}
            />
            <Button style={buttonStyles.buttonText} containerStyle={buttonStyles.button} onPress={this.onPress.bind(this)}>
              Submit
            </Button>
          </View>
        </ScrollView>
      </View>
    );
  }
}

SCForm.propTypes = {
  formInfo: PropTypes.object.isRequired
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: 0,
    justifyContent: 'center',
    flexDirection: 'column',
  },
  success: {
    flex: 1,
    padding: 10,
    backgroundColor: palette.gray
  },
  scrollView: {
    flex: 1,
  },
  formName: {
    backgroundColor: 'white',
    paddingTop: 5,
    paddingBottom: 5,
    paddingLeft: 20,
    paddingRight: 20,
    borderColor: '#bbb',
    borderBottomWidth: 1
  },
  formNameText: {
    color: '#333',
    fontSize: 24
  },
  form: {
    padding: 20
  }
});

export default SCForm;
