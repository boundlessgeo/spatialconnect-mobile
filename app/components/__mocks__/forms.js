export default [{
  id: 1,
  version: 0,
  form_key: 'weed_inspector',
  form_label: 'Weed Inspector',
  fields: [{
    id: 1,
    type: 'number',
    field_label: 'Weed Stage',
    field_key: 'weed_stage',
    is_required: true,
    position: 0,
    is_integer: false
  }, {
    id: 2,
    type: 'counter',
    field_label: 'Number of Plants',
    field_key: 'number_of_plants',
    position: 1,
    initial_value: '0'
  }, {
    id: 3,
    type: 'slider',
    field_label: 'Infestation Size',
    field_key: 'infestation_size',
    position: 2,
    initial_value: '0',
    minimum: '0',
    maximum: '100'
  }, {
    id: 4,
    type: 'counter',
    field_label: 'Degree',
    field_key: 'degree',
    is_required: true,
    position: 3,
    initial_value: '0',
    minimum: '0'
  }, {
    id: 5,
    type: 'slider',
    field_label: 'Density',
    field_key: 'density',
    position: 4,
    initial_value: '0',
    minimum: '0',
    maximum: '100'
  }, {
    id: 6,
    type: 'string',
    field_label: 'Source',
    field_key: 'source',
    is_required: true,
    position: 5,
    pattern: ''
  }, {
    id: 7,
    type: 'date',
    field_label: 'Data Discovered',
    field_key: 'date_discovered',
    position: 6
  }, {
    id: 8,
    type: 'string',
    field_label: 'Common Name',
    field_key: 'common_name',
    is_required: true,
    position: 7,
    pattern: ''
  }, {
    id: 9,
    type: 'string',
    field_label: 'Genus',
    field_key: 'genus',
    is_required: true,
    position: 8,
    pattern: ''
  }, {
    id: 10,
    type: 'string',
    field_label: 'Species',
    field_key: 'species',
    position: 9,
    pattern: ''
  }, {
    id: 11,
    type: 'string',
    field_label: 'Scientific Name',
    field_key: 'scientific_name',
    position: 10,
    pattern: ''
  }, {
    id: 12,
    type: 'string',
    field_label: 'Class',
    field_key: 'class',
    position: 11,
    pattern: ''
  }]
}, {
  id: 2,
  version: 0,
  form_key: 'baseball_team',
  form_label: 'Baseball Team',
  fields: [{
    id: 13,
    type: 'string',
    field_label: 'Favorite?',
    field_key: 'team',
    position: 0
  }, {
    id: 14,
    type: 'string',
    field_label: 'Why?',
    field_key: 'why',
    position: 1
  }]
}, {
  id: 5,
  version: 3,
  form_key: 'bird_counter',
  form_label: 'Bird Counter',
  fields: [{
    id: 18,
    type: 'counter',
    field_label: 'Bird Count',
    field_key: 'bird_count',
    is_required: true,
    position: 1,
    initial_value: '0'
  }, {
    id: 17,
    type: 'string',
    field_label: 'Bird Species',
    field_key: 'bird_species',
    is_required: true,
    position: 0,
    pattern: ''
  }]
}];