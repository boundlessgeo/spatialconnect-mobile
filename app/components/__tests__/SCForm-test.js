/*global jest,describe,it,expect*/
import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import SCForm from '../SCForm';
import forms from 'forms';

describe('<SCForm />', () => {

  it('renders correctly - Baseball Team', () => {
    const tree = renderer.create(
      <SCForm formInfo={forms[1]}/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

});