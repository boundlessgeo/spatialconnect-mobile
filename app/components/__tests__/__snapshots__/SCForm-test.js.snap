exports[`<SCForm /> renders correctly - Baseball Team 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
      "marginTop": 0
    }
  }>
  <ScrollView
    style={
      Object {
        "flex": 1
      }
    }>
    <View
      style={
        Object {
          "backgroundColor": "white",
          "borderBottomWidth": 1,
          "borderColor": "#bbb",
          "paddingBottom": 5,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 5
        }
      }>
      <Text
        style={
          Object {
            "color": "#333",
            "fontSize": 24
          }
        }>
        Baseball Team
      </Text>
    </View>
    <View
      style={
        Object {
          "padding": 20
        }
      }>
      <View
        style={Object {}}>
        <View
          style={
            Object {
              "marginBottom": 10
            }
          }>
          <Text
            style={
              Object {
                "color": "#000000",
                "fontSize": 17,
                "fontWeight": "500",
                "marginBottom": 7
              }
            }>
            Favorite?
          </Text>
          <TextInput
            accessibilityLabel="Favorite?"
            autoCapitalize={undefined}
            autoCorrect={undefined}
            autoFocus={undefined}
            blurOnSubmit={undefined}
            clearButtonMode={undefined}
            clearTextOnFocus={undefined}
            editable={undefined}
            enablesReturnKeyAutomatically={undefined}
            keyboardAppearance={undefined}
            keyboardType={undefined}
            maxLength={undefined}
            multiline={undefined}
            numberOfLines={undefined}
            onBlur={undefined}
            onChange={undefined}
            onChangeText={[Function onChangeText]}
            onEndEditing={undefined}
            onFocus={undefined}
            onKeyPress={undefined}
            onLayout={undefined}
            onSelectionChange={undefined}
            onSubmitEditing={undefined}
            placeholder={undefined}
            placeholderTextColor={undefined}
            returnKeyType={undefined}
            secureTextEntry={undefined}
            selectTextOnFocus={undefined}
            selectionColor={undefined}
            selectionState={undefined}
            style={
              Object {
                "backgroundColor": "#ffffff",
                "borderColor": "#cccccc",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#000000",
                "fontSize": 17,
                "height": 36,
                "marginBottom": 5,
                "padding": 7
              }
            }
            underlineColorAndroid={undefined}
            value="" />
        </View>
        <View
          style={
            Object {
              "marginBottom": 10
            }
          }>
          <Text
            style={
              Object {
                "color": "#000000",
                "fontSize": 17,
                "fontWeight": "500",
                "marginBottom": 7
              }
            }>
            Why?
          </Text>
          <TextInput
            accessibilityLabel="Why?"
            autoCapitalize={undefined}
            autoCorrect={undefined}
            autoFocus={undefined}
            blurOnSubmit={undefined}
            clearButtonMode={undefined}
            clearTextOnFocus={undefined}
            editable={undefined}
            enablesReturnKeyAutomatically={undefined}
            keyboardAppearance={undefined}
            keyboardType={undefined}
            maxLength={undefined}
            multiline={undefined}
            numberOfLines={undefined}
            onBlur={undefined}
            onChange={undefined}
            onChangeText={[Function onChangeText]}
            onEndEditing={undefined}
            onFocus={undefined}
            onKeyPress={undefined}
            onLayout={undefined}
            onSelectionChange={undefined}
            onSubmitEditing={undefined}
            placeholder={undefined}
            placeholderTextColor={undefined}
            returnKeyType={undefined}
            secureTextEntry={undefined}
            selectTextOnFocus={undefined}
            selectionColor={undefined}
            selectionState={undefined}
            style={
              Object {
                "backgroundColor": "#ffffff",
                "borderColor": "#cccccc",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#000000",
                "fontSize": 17,
                "height": 36,
                "marginBottom": 5,
                "padding": 7
              }
            }
            underlineColorAndroid={undefined}
            value="" />
        </View>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        onLayout={undefined}
        onResponderGrant={[Function bound touchableHandleResponderGrant]}
        onResponderMove={[Function bound touchableHandleResponderMove]}
        onResponderRelease={[Function bound touchableHandleResponderRelease]}
        onResponderTerminate={[Function bound touchableHandleResponderTerminate]}
        onResponderTerminationRequest={[Function bound touchableHandleResponderTerminationRequest]}
        onStartShouldSetResponder={[Function bound touchableHandleStartShouldSetResponder]}
        style={
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "#35507D",
            "borderColor": "#35507D",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 36,
            "justifyContent": "center",
            "margin": 0,
            "marginBottom": 10,
            "opacity": 1,
            "padding": 5
          }
        }
        testID={undefined}>
        <Text
          style={
            Array [
              Object {
                "color": "#007aff",
                "fontFamily": ".HelveticaNeueInterface-MediumP4",
                "fontSize": 17,
                "fontWeight": "bold",
                "textAlign": "center"
              },
              null,
              Object {
                "alignSelf": "center",
                "color": "white",
                "fontSize": 16
              },
              null
            ]
          }>
          Submit
        </Text>
      </View>
    </View>
  </ScrollView>
</View>
`;
